apps/src/code-studio/initApp/project.js

```diff
@@ -960,6 +963,38 @@ var projects = (module.exports = {
       });
     });
   },
+
+  /**
+   * Tests whether provided sample code is different from the current project code.
+   * This also normalizes the code so incidental differences in line endings or
+   * empty xml tags are not recognized as differences.
+   * @param {string} sampleCode the code to diff against the current project code
+   */
+  isCodeDifferent(sampleCode) {
+    let normalizedSample, normalizedCurrent;
+    const parser = new DOMParser();
+    const parsedCurrent = parser.parseFromString(
+      currentSources.source,
+      'text/xml'
+    );
+    const parsedSample = parser.parseFromString(sampleCode, 'text/xml');
+    if (
+      parsedCurrent.getElementsByTagName('parsererror').length > 0 ||
+      parsedSample.getElementsByTagName('parsererror').length > 0
+    ) {
```

Comment:
Why does this check help determine if it's blockly or droplet?

Reply:
More specifically, it determines if the text is xml or plaintext. For our product, that translates to "blockly or droplet" 
^ Does that answer your question?

Reply:
Yes, that answers my question.  I think it's a little bit confusing.  One idea might be to extract a function that returns whether the given sample is blockly or droplet.  Or is there a property that more explicitly identifies a project as blockly or droplet?  I'm also ok to leave as-is that sounds like too much trouble.  Up to you.

Reply:
Oh hmm. I hadn't thought about it that way. I was approaching this from the direction of "how to diff xml and diff text" rather than "how to diff blockly and diff droplet" although for our purposes, those are equivalent. It seems a little odd that project.js would know whether an app is using blockly or droplet as it's supposed to just handle saving/loading/etc the project's data and metadata. But it looks like it does have access to that information from `appOptions.droplet`. 

Another thought: There is theoretically the possibility that blockly code could somehow cause a parseError if the xml data got malformed somehow. In that case, we wouldn't want to naively diff the serialized xml and would use the text diff as a fallback. But I'm guessing a parseError in the blockly xml would break a lot of other things too and we likely wouldn't even get to this point.

Oof sorry for the block of text. I was typing while I was thinking.

Reply:
Ohhh, I see, that makes sense now.  I didn't realize that project.js is agnostic to the project type.  And I think the comment right above the "if" check made me think that the intent was more to differentiate blockly code from droplet code.  I'm fine with this now, thanks for taking the time to explain it to me.